stl/inc/xutility

```diff
@@ -4889,23 +4889,120 @@ public:
+        // For the __shiftright128 intrinsic, the shift value is always
+        // modulo 64.
+        // In the current implementation of the double-precision version
+        // of Ryu, the shift value is always < 64.
+        // (The shift value is in the range [49, 58].)
+        // Check this here in case a future change requires larger shift
+        // values. In this case this function needs to be adjusted.
+        _STL_INTERNAL_CHECK(__dist < 64);
+        return __shiftright128(__lo, __hi, static_cast<unsigned char>(__dist));
+    }
+
+#else // ^^^ intrinsics available ^^^ / vvv intrinsics unavailable vvv
+
+    _NODISCARD static __forceinline uint64_t __rng_umul128(
+        const uint64_t __a, const uint64_t __b, uint64_t* const __productHi) {
+        // TRANSITION, VSO-634761
+        // The casts here help MSVC to avoid calls to the __allmul library function.
+        const uint32_t __aLo = static_cast<uint32_t>(__a);
+        const uint32_t __aHi = static_cast<uint32_t>(__a >> 32);
+        const uint32_t __bLo = static_cast<uint32_t>(__b);
+        const uint32_t __bHi = static_cast<uint32_t>(__b >> 32);
+
+        const uint64_t __b00 = static_cast<uint64_t>(__aLo) * __bLo;
+        const uint64_t __b01 = static_cast<uint64_t>(__aLo) * __bHi;
+        const uint64_t __b10 = static_cast<uint64_t>(__aHi) * __bLo;
+        const uint64_t __b11 = static_cast<uint64_t>(__aHi) * __bHi;
+
+        const uint32_t __b00Lo = static_cast<uint32_t>(__b00);
+        const uint32_t __b00Hi = static_cast<uint32_t>(__b00 >> 32);
+
+        const uint64_t __mid1   = __b10 + __b00Hi;
+        const uint32_t __mid1Lo = static_cast<uint32_t>(__mid1);
+        const uint32_t __mid1Hi = static_cast<uint32_t>(__mid1 >> 32);
+
+        const uint64_t __mid2   = __b01 + __mid1Lo;
+        const uint32_t __mid2Lo = static_cast<uint32_t>(__mid2);
+        const uint32_t __mid2Hi = static_cast<uint32_t>(__mid2 >> 32);
+
+        const uint64_t __pHi = __b11 + __mid1Hi + __mid2Hi;
+        const uint64_t __pLo = (static_cast<uint64_t>(__mid2Lo) << 32) | __b00Lo;
+
+        *__productHi = __pHi;
+        return __pLo;
+    }
+
+    _NODISCARD static inline uint64_t __rng_shiftright128(
+        const uint64_t __lo, const uint64_t __hi, const uint32_t __dist) {
+        // We don't need to handle the case __dist >= 64 here (see above).
+        _STL_INTERNAL_CHECK(__dist < 64);
+#ifdef _WIN64
+        _STL_INTERNAL_CHECK(__dist > 0);
+        return (__hi << (64 - __dist)) | (__lo >> __dist);
+#else // ^^^ 64-bit ^^^ / vvv 32-bit vvv
+      // Avoid a 64-bit shift by taking advantage of the range of shift values.
+        _STL_INTERNAL_CHECK(__dist >= 32);
+        return (__hi << (64 - __dist)) | (static_cast<uint32_t>(__lo >> 32) >> (__dist - 32));
+#endif // ^^^ 32-bit ^^^
+    }
+
+#endif // ^^^ intrinsics unavailable ^^^
+
+    // using multiply-shift instead of modulo to fairly map values in range [0, _Mod), adapted from arXiv:1805.10941v4

```

Comment:
We'll need to check if this is OK. Did you refer to only the blog post and the PDF on arXiv, but not the `fastrange` repository?

Reply:
Well, that's embarrassing. I actually missed the github link to the standalone header, as I figured that would also refer to the file linked to from "As usual, my code is freely available."

Reply:
Thanks; I'm looking into this now.

Reply:
We're good to go! I'm busy with other things this week, but I should be able to update your PR soon.